### PR TITLE
Add condition and index for join

### PR DESCRIPTION
**Pull request by pierre-H**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | none

This is useful for custom filters.

### CODE REVIEW

**Comment by soyuka** [comment]
Interesting, can you give us an example use case?

Note that sometimes, the filter doesn't join, it's done in the EagerLoadingExtension instead.

**Comment by pierre-H** [comment]
I've got an entity called "User" which have a relation OneToMany to the "Stay" entity.
In the Stay entity I've got a "state" field (no started, in progress, done) and a "date" field.

I need to filter Users with the state field of the last Stay.

Example : 

Users : 
1.  Marc
2.  Pierre

Stay : 
1. UserId: 1 - Date : 10/01/16 - State : In progress
2. UsrId : 2 - Date : 14/02/2016 - State : In progress
3. UsrId : 1 - Date : 23/02/2016 - State : No started

So, with a field : "State", if I select : "In progress" I will get only Pierre and no Marc because his last Stay's state is : "No started".

For this I need to perform this join : 

```
$queryBuilder->join(
	'o.stays',
	'stay',
	'WITH',
	'( o.id= stay.userId AND stay.id= ( SELECT MAX (stay2.id) FROM AppBundle\Entity\Stay as stay2 WHERE stay2.userId = o.id) )');
```

---
Note : thank you so much for your great work !!! Api-Platform is very useful and impressive !

**Comment by soyuka** [comment]
So you extended the AbstractFilter to create your own? 
Did you disable the eager extension before using this, because if it's joined there (it should by default if `stays` is serialized) the filter will not do an extra join.

Btw you misspelled "string":

```
sting $indexBy = null
```

File: src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php [hunk 1]
```diff
@@ -219,20 +219,22 @@ protected function extractProperties(Request $request): array
...
+     * @param string                      $indexBy            the index association (see $queryBuilder->join() doc)
      *
      * @throws InvalidArgumentException If property is not nested
      *
      * @return array An array where the first element is the join $alias of the leaf entity,
      *               the second element is the $field name
      *               the third element is the $associations array
      */
-    protected function addJoinsForNestedProperty(string $property, string $rootAlias, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator): array
+    protected function addJoinsForNestedProperty(string $property, string $rootAlias, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $conditionType = null, sting $indexBy = null): array
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php]: string is misspelled

**Comment by dunglas** [comment]
We really need to improve docs about custom filters!

**Comment by soyuka** [comment]
> We really need to improve docs about custom filters!

Would you recommend to extend the AbstractFilter class? 

Personally I'm prefering data providers over filters, because if want I want isn't possible with built-in filters it's usually that I want a custom query, and therefore need to do way more work than just filtering. I mean, filter should be restraint to simple `WHERE` clauses, whereas a data provider is useful when you need custom `JOIN` etc.

**Comment by dunglas** [comment]
@soyuka, no it would be nice to have a simple example just implementing the `FilterInterface`.

File: src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php [hunk 2]
```diff
@@ -274,17 +276,19 @@ private function getExistingJoin(QueryBuilder $queryBuilder, string $alias, stri
      * @param QueryNameGeneratorInterface $queryNameGenerator
      * @param string                      $alias
      * @param string                      $association        the association field
+     * @param string                      $conditionType      the condition association (see $queryBuilder->join() doc)
+     * @param string                      $indexBy            the index association (see $queryBuilder->join() doc)
      *
      * @return string the new association alias
      */
-    protected function addJoinOnce(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $association): string
+    protected function addJoinOnce(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $association, string $conditionType = null, sting $indexBy = null): string
```
Comment (Simperfit) [inline src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php]: string misspelled too

**Comment by soyuka** [comment]
@Simperfit I think this PR should not be merged but we should rather improve the documentation on custom filters :)

**Comment by pierre-H** [comment]
@soyuka That's a better idea !

**Comment by soyuka** [comment]
Closing in favor of #922

**Pull request closed without merging**